src/pcp.rs

```diff
@@ -0,0 +1,596 @@
+
+    /// Returns an instance of the gadget associated with the validity circuit. `in_len` is the
+    /// maximum degree of the each polynomial input to `call_poly`. If this function is never used
+    /// by the caller, then it is save to set `in_len == 0`.
+    fn gadget(&self, in_len: usize) -> G;
+}
+
+/// The gadget functionality required for evaluating a validity circuit. The `Gadget` trait
+/// inherits this trait.
+pub trait GadgetCallOnly<F: FieldElement> {
+    /// Evaluates the gadget on input `inp` and returns the output.
+    fn call(&mut self, inp: &[F]) -> Result<F, PcpError>;
+
+    /// Returns the circuit's arity, i.e., the expected length of the input to `call`.
+    fn call_in_len(&self) -> usize;
+}
+
+/// The sub-circuit associated with some validity circuit. A gadget is called either on a seqeuence
+/// of finite field elements or a sequence of polynomials over a finite field.
+pub trait Gadget<F: FieldElement>: GadgetCallOnly<F> {
+    /// Evaluate the gaget on input of a sequence of polynomials. The output is written to `outp`.
+    fn call_poly<V: AsRef<[F]>>(&mut self, outp: &mut [F], inp: &[V]) -> Result<(), PcpError>;
+
+    /// The maximum degree of the polynomial output by `call_poly` as a function of the maximum
+    /// degree of each input polynomial.
+    fn call_poly_out_len(&self, in_len: usize) -> usize;
+}
+
+/// Generate a proof of an input's validity.
+pub fn prove<F, G, T>(x: &T) -> Result<Proof<F>, PcpError>
+where
+    F: FieldElement,
+    G: Gadget<F>,
+    T: Type<F, G>,
+{
+    let g_calls = x.valid_gadget_calls();
+    let m = (g_calls + 1).next_power_of_two();
+    let mut g = x.gadget(m);
+    let p = g.call_poly_out_len(m);
+    let l = g.call_in_len();
+    let mut data = vec![F::zero(); l + p];
+
+    // Run the validity crciut with a "shim" gadget that records the value of each input wire of
+    // each gadget evaluation.
+    let mut shim = ProveShimGadget::new(&mut g, g_calls);
+    let _ = x.valid(&mut shim, &rand_vec(x.valid_rand_len()));
+
+    // Construct the intermediate proof polynomials `f[0], ..., f[l-1]`. Also, record in the slice
+    // `data[...l]` the value of `f[i](1)`, i.e., the first point at which polynomial `f[i]` was
+    // interpolated.
+    let mut f = vec![vec![F::zero(); m]; l];
+    let m_inv = F::from(F::Integer::try_from(m).unwrap()).inv();
+    for i in 0..l {
+        data[i] = shim.f_vals[i][0];
+        discrete_fourier_transform(&mut f[i], &shim.f_vals[i], m)?;
+        discrete_fourier_transform_inv_finish(&mut f[i], m, m_inv);
+    }
+
+    // Construct the proof polynomial `data[l..] = G(f[0], ..., f[l-1])`.
+    g.call_poly(&mut data[l..], &f)?;
+
+    let poly_len = poly_deg(&data[l..]);
+    data.truncate(l + poly_len + 1);
+    Ok(Proof { data })
+}
+
+// A "shim" gadget used during proof generation to record the points at which the intermediate
+// proof polynomials are interpolated.
+struct ProveShimGadget<'a, F, G>
```

Comment:
My weak preference would be to leave it where it is, mainly because it's specific to the implementation of `query`, whereas everything in `gadgets.rs` are meant to be public. I don't feel that strongly, so feel free to override me.